File: everything_parsing/28-grammar-tool-project2/js/grammartool.test.mjs
// node js/grammartool.test.mjs

import { that, equals, summary } from "./assert.mjs";
import { Grammar } from "./grammar.mjs";
import { GrammarTool } from "./grammartool.mjs";
import { G, paull, hasLeftRecursion } from "./leftrec.mjs";
import { factor, needsFactoring } from "./leftfactor.mjs";
import { Predict } from "./predict.mjs";

console.log("grammartool.test.mjs");

const s = set => JSON.stringify([...set].sort());
const toGrammar = lg => Grammar.parse("%start " + lg.start + "\n" + lg.text());

// ── LL(1) statement grammar ──
const st = Grammar.parse(`
%start P
P -> L
L -> S L | epsilon
S -> id assign E semi
E -> id | num
`);
const t1 = new GrammarTool(st);
equals([...st.nonterminals].join(" "), "P L S E", "nonterminals in appearance order");
equals(s(t1.fs.nullable), s(new Set(["L", "P"])), "L and P nullable");
equals(s(t1.fs.firstOf("E")), s(new Set(["id", "num"])), "FIRST(E)");
equals(s(t1.fol.followOf("L")), s(new Set(["$"])), "FOLLOW(L)");
that(new Predict(st).isLL1(), "already LL(1)");
that(!hasLeftRecursion(paull(st)), "Paull leaves it LL-friendly");
that(!needsFactoring(factor(st)), "nothing to factor");

// ── natural expression grammar ──
const ex = Grammar.parse(`
%start E
E -> E + T | T
T -> T * F | F
F -> ( E ) | id | num
`);
that(!new Predict(ex).isLL1(), "expression grammar is not LL(1)");
that(hasLeftRecursion(G.from(ex)), "it is left-recursive");

const noLR = paull(ex);
equals(noLR.text().trim(), [
  "E -> T E'",
  "E' -> + T E' | epsilon",
  "T -> F T'",
  "T' -> * F T' | epsilon",
  "F -> ( E ) | id | num",
].join("\n"), "section 5 gives the Module 19 grammar");
that(new Predict(toGrammar(noLR)).isLL1(), "and that IS LL(1)");

// ── JSON grammar ──
const js = Grammar.parse(`
%start value
value    -> obj | arr | str | num | tru | fls | nul
obj      -> lbrace rbrace | lbrace members rbrace
members  -> pair | pair comma members
pair     -> str colon value
arr      -> lbrack rbrack | lbrack elements rbrack
elements -> value | value comma elements
`);
that(!new Predict(js).isLL1(), "JSON grammar as written is not LL(1)");
that(!hasLeftRecursion(G.from(js)), "but it has no left recursion");
that(paull(js).text().includes("elements -> value | value comma elements"),
  "Paull leaves 'elements' untouched");
that(!needsFactoring(factor(js)), "left factoring clears the shared prefixes");
that(new Predict(toGrammar(factor(toGrammar(paull(js))))).isLL1(), "transformed JSON grammar is LL(1)");

// ── report runs end to end ──
const rep = new GrammarTool(ex).report();
that(rep.includes("7. LL(1) VERDICT"), "report has all sections");
that(rep.includes("after left-recursion removal + left factoring: LL(1) = YES"),
  "report confirms the fix works");

summary();

File: everything_parsing/28-grammar-tool-project2/java/GrammarToolTest.java
import java.util.Set;
import java.util.TreeSet;

/** java -cp java/out GrammarToolTest */
public final class GrammarToolTest {

    static String s(Set<String> set) { return new TreeSet<>(set).toString(); }

    public static void main(String[] args) {
        System.out.println("GrammarToolTest");

        // ── the LL(1) statement grammar: all seven sections agree it's ready ──
        Grammar st = Grammar.parse("""
            %start P
            P -> L
            L -> S L | epsilon
            S -> id assign E semi
            E -> id | num
            """);
        GrammarTool t1 = new GrammarTool(st);
        Assert.equals(String.join(" ", st.nonterminals), "P L S E", "nonterminals in appearance order");
        Assert.equals(s(t1.fs.nullable), "[L, P]", "L and P nullable");
        Assert.equals(s(t1.fs.firstOf("E")), "[id, num]", "FIRST(E)");
        Assert.equals(s(t1.fol.followOf("L")), "[$]", "FOLLOW(L)");
        Assert.that(new Predict(st).isLL1(), "already LL(1)");
        Assert.that(!LeftRec.hasLeftRecursion(LeftRec.paull(st)), "Paull leaves it LL-friendly");
        Assert.that(!LeftFactor.needsFactoring(LeftFactor.factor(st)), "nothing to factor");

        // ── the natural expression grammar: NOT LL(1), fixable ──
        Grammar ex = Grammar.parse("""
            %start E
            E -> E + T | T
            T -> T * F | F
            F -> ( E ) | id | num
            """);
        GrammarTool t2 = new GrammarTool(ex);
        Assert.that(!new Predict(ex).isLL1(), "expression grammar is not LL(1)");
        Assert.that(LeftRec.hasLeftRecursion(LeftRec.G.from(ex)), "it is left-recursive");

        LeftRec.G noLR = LeftRec.paull(ex);
        Assert.equals(noLR.text().strip(),
            String.join("\n",
                "E -> T E'",
                "E' -> + T E' | epsilon",
                "T -> F T'",
                "T' -> * F T' | epsilon",
                "F -> ( E ) | id | num"),
            "section 5 gives the Module 19 grammar");
        Grammar fixed = GrammarTool.toGrammar(noLR);
        Assert.that(new Predict(fixed).isLL1(), "and that IS LL(1)");

        // ── the JSON grammar: not LL(1) (shared prefixes), left factoring fixes it ──
        Grammar js = Grammar.parse("""
            %start value
            value    -> obj | arr | str | num | tru | fls | nul
            obj      -> lbrace rbrace | lbrace members rbrace
            members  -> pair | pair comma members
            pair     -> str colon value
            arr      -> lbrack rbrack | lbrack elements rbrack
            elements -> value | value comma elements
            """);
        Assert.that(!new Predict(js).isLL1(), "JSON grammar as written is not LL(1)");
        Assert.that(!LeftRec.hasLeftRecursion(LeftRec.G.from(js)), "but it has no left recursion");
        // section 5 must NOT mangle the non-left-recursive elements rule
        LeftRec.G jsNoLR = LeftRec.paull(js);
        Assert.that(jsNoLR.text().contains("elements -> value | value comma elements"),
            "Paull leaves 'elements' untouched (no left recursion to remove)");
        LeftRec.G jsFactored = LeftFactor.factor(js);
        Assert.that(!LeftFactor.needsFactoring(jsFactored), "left factoring clears the shared prefixes");
        Grammar jsFixed = GrammarTool.toGrammar(LeftFactor.factor(GrammarTool.toGrammar(LeftRec.paull(js))));
        Assert.that(new Predict(jsFixed).isLL1(), "transformed JSON grammar is LL(1)");

        // ── the report runs end to end ──
        Assert.that(t2.report().contains("7. LL(1) VERDICT"), "report has all sections");
        Assert.that(t2.report().contains("after left-recursion removal + left factoring: LL(1) = YES"),
            "report confirms the fix works");

        Assert.summary();
    }
}
